File: solutions/cpp/0017.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};

        vector<string> ret;

        unordered_map<char, string> map = {
            {'2', "abc"}, {'3', "def"},  {'4', "ghi"}, {'5', "jkl"},
            {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};

        ret.push_back("");
        for (const char& i : digits) {
            vector<string> tmp;
            for (const string& j : ret)
                for (const char& k : map[i])
                    tmp.push_back(j + k);
            ret = tmp;
        }

        return ret;
    }
};

File: solutions/python3/0017.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        ans = ['']

        map = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
               '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}

        for i in digits:
            tmp = []
            for j in ans:
                for k in map[i]:
                    tmp.append(j + k)
            ans = tmp

        return ans
